### PR TITLE
Update yara rules and check any new rules against existing files

### PR DESCRIPTION
**Pull request by MancunianSam**
The steps for this are roughly.
* Get the old rules file from S3
* Build the rules docker image to get an updated one. 
* Run the new python script against the two files which compares them.
* If there are new rules in the new compiled rules, the new rules are run against the test files in the S3 bucket.

### CODE REVIEW

File: Dockerfile-run-tests [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FROM yara-rules
```
Comment (TomJKing) [inline Dockerfile-run-tests]: Maybe its worth renaming this Docker file? Might be a bit confusing as could mean run the unit tests? Maybe something like: `Dockerfile-check-rules`

File: Jenkinsfile-check-rules [hunk 2]
```diff
@@ -0,0 +1,57 @@
+library("tdr-jenkinslib")
+
+def yaraVersion = "4.1.3"
```
Comment (TomJKing) [inline Jenkinsfile-check-rules]: Will this value need updating if a newer version of yara is released?

File: Jenkinsfile-check-rules [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+pipeline {
+  agent {
+    label "built-in"
+  }
+  stages {
+    stage("Check for updates") {
```
Comment (TomJKing) [inline Jenkinsfile-check-rules]: Maybe rename stage: `Check for yara rules updates`

File: Jenkinsfile-test [hunk 4]
```diff
@@ -2,7 +2,7 @@ library("tdr-jenkinslib")
 
 def repo = "tdr-antivirus"
 def versionTag = "v${env.BUILD_NUMBER}"
-def yaraVersion = "4.1.0"
+def yaraVersion = "4.1.3"
```
Comment (TomJKing) [inline Jenkinsfile-test]: Is there anyway to define the yara version in a single place? Its being defined in multiple files.

File: README.md [hunk 5]
```diff
@@ -79,3 +79,22 @@ Run with coverage `python -m pytest --cov=src`
 Run with coverage and missing lines `python -m pytest --cov-report term-missing --cov=src`
 
 Run with coverage, missing lines and junit output `python -m pytest --cov-report term-missing --junitxml="result.xml" --cov=src`
+
+## Yara rules checks.
+
```
Comment (TomJKing) [inline README.md:84]: Is it worth adding some instructions for the Terraform. It's not going to need deploying very much, but it still needs covering in the README I think.

File: README.md [hunk 6]
```diff
@@ -79,3 +79,22 @@ Run with coverage `python -m pytest --cov=src`
...
+There is a Jenkins job run on a schedule from the Jenkinsfile-check-rules file. This carries out the following steps.
```
Comment (TomJKing) [inline README.md]: Maybe reference the name of the Jenkins job.

File: README.md [hunk 7]
```diff
@@ -79,3 +79,22 @@ Run with coverage `python -m pytest --cov=src`
 Run with coverage and missing lines `python -m pytest --cov-report term-missing --cov=src`
 
 Run with coverage, missing lines and junit output `python -m pytest --cov-report term-missing --junitxml="result.xml" --cov=src`
+
+## Yara rules checks.
+
+There is a Jenkins job run on a schedule from the Jenkinsfile-check-rules file. This carries out the following steps. 
+* Builds the base yara image.
+* Builds the rules yara image.
+* Gets the most recent version from git.
```
Comment (TomJKing) [inline README.md]: Maybe reword to make clear what the "recent version" is?

File: README.md [hunk 8]
```diff
@@ -79,3 +79,22 @@ Run with coverage `python -m pytest --cov=src`
...
+* Downloads the lambda zip file from S3.
+* Unzips the zip file and copies the exiting compiled rule file into the current directory.
+* Downloads the test files from tdr-antivirus-test-files-mgmt in S3.
```
Comment (TomJKing) [inline README.md]: Maybe reword slightly: "Downloads the tests file from the tdr-antivirus-test-files-mgmt S3 bucket"

File: README.md [hunk 9]
```diff
@@ -79,3 +79,22 @@ Run with coverage `python -m pytest --cov=src`
...
+* Builds the docker image from the Dockerfile-run-tests file which copies the existing compiled rules and the test files to the container.
+* Runs this docker image. This runs a python script which runs these steps:
+    * Compares the rule identifiers in the old compiled rules file with the new one.
+    * If there are new rules in the new compiled rules, it will run the checks against the files. It will not detect if rules have been removed and it won't check if they've been changed either. I tried to do it based on hashes but it's likely to bring up too many false positives.
```
Comment (TomJKing) [inline README.md]: Maybe reword this: 

> If there are new yara rules in the new compiled rules, it will run the antivirus checks against the test files from the S3 bucket. 
> * NOTE: It will not detect if rules have been removed, and it won't check if they've been changed either. I tried to do it based on hashes but it's likely to bring up too many false positives.

File: README.md [hunk 10]
```diff
@@ -79,3 +79,22 @@ Run with coverage `python -m pytest --cov=src`
...
+    * Yara is run against the test files using the new rules. If it finds no matches, if returns exit code zero and the Jenkins job builds the Antivirus multibranch master branch job.
```
Comment (TomJKing) [inline README.md]: Maybe reword slight: "If no matches are found after the new rules are run against the test files, it returns exit code zero ..."

File: terraform/root.tf [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+provider "aws" {
+  region = local.aws_region
+}
+
+variable "default_aws_region" {
+  default = "eu-west-2"
+}
+
+terraform {
```
Comment (TomJKing) [inline terraform/root.tf]: Can you split the `terraform {... }` from line 27 to 42 into a versions.tf file, just to keep it consistent with the Terraform structure in all the other repos?

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor comments. I might misunderstood some the steps, so some of the reword suggestions might be incorrect.

File: Dockerfile-run-tests (hunk 1 above)
Comment (MancunianSam) [reply Dockerfile-run-tests]: Yeah, makes sense. I've renamed it and updated the Jenkinsfile.

File: Jenkinsfile-check-rules (hunk 2 above)
Comment (MancunianSam) [reply Jenkinsfile-check-rules]: Yep, that's the case now for the main antivirus build as well. I think we thought about having it as a Jenkins parameter but they don't release a new version that often so we just hard coded it.

File: Jenkinsfile-test (hunk 4 above)
Comment (MancunianSam) [reply Jenkinsfile-test]: I've added in a yara-version.txt file which both of the Jenkinsfiles read from and this looks like it's working so we'd only have to change one file. 

The other place it's in is in the requirements txt file for yara-python. I can't see a way of reading this globally. It's the version of the python library though and not the version of yara itself so I think it's ok to keep it separate, although they should be kept in sync really
The other thing to do with the python version is just to remove the version and always use the latest but that might cause  other problems down the line.

File: README.md (hunk 5 above)
Comment (MancunianSam) [reply README.md:84]: I've put a couple of lines in saying what its for.

File: README.md (hunk 6 above)
Comment (MancunianSam) [reply README.md]: I've put the name in and a link to the job.

File: README.md (hunk 7 above)
Comment (MancunianSam) [reply README.md]: I've put "highest version tag" which is a bit more descriptive.

File: README.md (hunk 9 above)
Comment (MancunianSam) [reply README.md]: Yeah that's better, I've changed it.

File: terraform/root.tf (hunk 11 above)
Comment (MancunianSam) [reply terraform/root.tf]: I've moved it out to the versions.tf file.

File: README.md [hunk 12]
```diff
@@ -79,3 +79,27 @@ Run with coverage `python -m pytest --cov=src`
...
 Run with coverage, missing lines and junit output `python -m pytest --cov-report term-missing --junitxml="result.xml" --cov=src`
+
+## Yara rules checks.
+
+There is a Jenkins job [TDR Antivirus Rule Update Check](https://jenkins.tdr-management.nationalarchives.gov.uk/job/TDR%20Antivirus%20Rule%20Update%20Check/) which is run on a schedule from the Jenkinsfile-check-rules file. This carries out the following steps. 
+* Builds the base yara image.
+* Builds the rules yara image.
+* Gets the highest version tag from git.
+* Downloads the lambda zip file from S3.
+* Unzips the zip file and copies the exiting compiled rule file into the current directory.
```
Comment (TomJKing) [inline README.md]: Spelling "exiting" should be "existing"

File: README.md [hunk 13]
```diff
@@ -79,3 +79,27 @@ Run with coverage `python -m pytest --cov=src`
...
+        * NOTE: It will not detect if rules have been removed, and it won't check if they've been changed either. I tried to do it based on hashes but it's likely to bring up too many false positives.
+    * Yara is run against the test files using the new rules. If no matches are found after the new rules are run against the test files, it returns exit code zero and the Jenkins job builds the Antivirus multibranch master branch job.
+  
+These steps can be run manually by copying the commands from the Jenkinsfile-check-rules.
+
+Because this runs the master build job without any code changes, you end up with more than one version pointed to the same commit. We still have separate zip files stored in S3 for each version though so we can still roll back if necessary.
+
+## Test files bucket terraform
+
+In the terraform directory are some terraform files which are used to create the bucket `tdr-antivirus-test-files-mgmt`. This bucket is used to store the files against which we run a periodic check of any new yara rules. This should almost never need to be updated.
```
Comment (TomJKing) [inline README.md]: Missing word: ... directory **there** are some ..

**Review by TomJKing — COMMENTED** [review]
Just a couple of minor typos to fix.

**Review by TomJKing — APPROVED** [review]
